### PR TITLE
[EI-519][indexer][testing] Implement scripted transaction generation

### PR DESCRIPTION
**Pull request by larry-aptos**
## Description

* Scripted transactions with a shared step

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [x] Other (specify)

## How Has This Been Tested?

Locally tested.

* Command used: 
```bash
cargo run -- --testing-folder /absolute/path/to/testing_folder --output-folder /your_output/folder
```

* Example testing folder structure
image

* Result
image

importing txns with this imported_transactions.yaml
```
testnet:
  # Transaction Stream endpoint addresss.
  transaction_stream_endpoint: https://grpc.testnet.aptoslabs.com:443
  # (Optional) The key to use with developers.aptoslabs.com
  api_key: aptoslabs_L7pzPsqiShp_EwzUCbhb6ehTe8PSvFp6ZeUr3B3gBdGaX
  # A map from versions to dump and their output names.
  versions_to_import:
    123: testnet_v1
    278556781: testnet_v1_coin_register_fa_metadata_278556781
    1255836496: testnet_v2_fa_metadata_1255836496
    5979639459: testnet_v5979639459
    5992795934: testnet_v5992795934
    51268254: testnet_v51268254


mainnet:
  # Transaction Stream endpoint addresss.
  transaction_stream_endpoint: https://grpc.testnet.aptoslabs.com:443
  # (Optional) The key to use with developers.aptoslabs.com
  api_key: aptoslabs_L7pzPsqiShp_EwzUCbhb6ehTe8PSvFp6ZeUr3B3gBdGaX
  # A map from versions to dump and their output names.
  versions_to_import:
    308783012: mainnet_v308783012


```

result
Screenshot 2024-09-27 at 12 41 46 PM

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by larry-aptos** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14569?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14569** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14569?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14551** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14551?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14550** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14550?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14549** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14549?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @larry-aptos and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md [hunk 1]
```diff
@@ -18,4 +19,11 @@ import_config:
     # A map from versions to dump and their output names.
     versions_to_import:
       123: testnet_v1.json
+# Config to generate the transactions via localnode.
+script_transaction_generator_config:
+  scripted_transactions:
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-transaction-generator/README.md:24]: How does the config handle generation of different transactions? Could add an example here to to demonstrate

Comment (rtso) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:24]: Also I remember we talked about renaming "steps" to "transactions" to be clarify what the output of this is. It's still not clear to me if a transaction has multiple steps, or if there's 1 transaction generated per step?

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md [hunk 2]
```diff
@@ -18,4 +19,11 @@ import_config:
...
+    # Steps can be shared between runs.
+    - steps:
+        - script_path: /path/to/your_move_script
+          output_name: random_script.json
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-transaction-generator/README.md:28]: Are there any other config variable options? Let's put them all here.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md [hunk 3]
```diff
@@ -9,6 +9,7 @@ This tool is to generate transactions for testing purpose.
 ### Config
 
 ```YAML
+# Config to import transactions onchain.
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/README.md:12]: I wonder how the config file would look like in prod, or when we have this Ci/CD with all transactions included.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    pub async fn run(&self, output_path: &Path) -> anyhow::Result<()> {
+        // Validate the configuration.
+        self.validate()?;
+
+        // Submit and capture the transactions.
+        for script_transactions in &self.scripted_transactions {
+            // Start the localnet.
+            // TODO: refresh the localnet for each script transaction.
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:57]: can you elaborate this

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            let mut managed_node = ManagedNode::start(None, None).await?;
+
+            let mut versions_to_capture = vec![];
+
+            for step in &script_transactions.steps {
+                let version = self.execute_script_step(step).await?;
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:63]: I think this might confuse people, let's rename this to transaction instead of step to be clear.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                .await?;
+
+            // Stop the localnet.
+            managed_node.stop().await?;
+        }
+        Ok(())
+    }
+
+    /// Prepare the script
+    async fn prepare_script_step(&self, step: &ScriptStep) -> anyhow::Result<()> {
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:78]: nit: rename this function name and add more context to the description such that we know why we are running this before other transaction(step)

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+};
+
+/// GRPC request metadata key for the token ID.
+const LOCAL_INDEXER_GRPC_URL: &str = "http://127.0.0.1:50051";
+const TRANSACTION_STREAM_TIMEOUT_IN_SECS: u64 = 60;
+const DEFAULT_FUND_AMOUNT_IN_OCTA: u64 = 100_000_000;
+
+impl ScriptTransactionGeneratorConfig {
+    fn validate(&self) -> Result<()> {
+        // Validate the script transactions.
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:31]: maybe we can also return early or raise an error if scripted_transactions is empty?

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        // 1. No output file names are duplicated.
+        let mut output_files = std::collections::HashSet::new();
+        for script_transactions in &self.scripted_transactions {
+            for step in &script_transactions.steps {
+                if let Some(output_name) = &step.output_name {
+                    if !output_files.insert(output_name) {
+                        return Err(anyhow::anyhow!(
+                            "[Script Transaction Generator] Output file name {} is duplicated",
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:39]: nit: we can provide more context like this

```                        "[Script Transaction Generator] Duplicate output file name '{}' found at transaction {} step {}.",```

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 1 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md]: done; updated.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 2 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md]: done. updated.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 3 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:12]: maybe under `indexer-test-transactions`? and one file looks like this, wdyt?

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs (hunk 4 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:57]: Oh currently for transactions with dependencies we put under the same run. if possible, we should move to single run for all transactions.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs (hunk 5 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs]: done.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs (hunk 6 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs]: done.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs (hunk 7 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:31]: done; error out when transactions are empty.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs (hunk 8 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:39]: updated. 
skipped the version; since the version may not be that useful considering it's localnet and we wipe it for each run..

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            .await
+            .context(format!("Failed to run the script: {:?}", path))?;
+        if let Some(true) = transaction_summary.success {
+            Ok(transaction_summary.version.unwrap())
+        } else {
+            anyhow::bail!("Failed to execute the script: {:?}", path);
+        }
+    }
+
+    async fn capture_transaction(
```
[the remark this answers is not in the document]
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs:146]: done.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        let _ = fund_cmd.execute().await.context(format!(
+            "Failed to fund the account for account: {}.",
+            step.fund_address
+                .clone()
+                .unwrap_or("Default profile".to_string())
+        ))?;
+        Ok(())
+    }
+
+    async fn execute_script_step(&self, step: &ScriptStep) -> anyhow::Result<u64> {
```
[the remark this answers is not in the document]
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/script_transaction_generator.rs]: done.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md [hunk 11]
```diff
@@ -18,4 +19,21 @@ import_config:
     # A map from versions to dump and their output names.
     versions_to_import:
       123: testnet_v1.json
+# Config to generate the transactions via localnode.
+script_transaction_generator_config:
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/README.md:23]: nit: can we give a placeholder config file?

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md [hunk 12]
```diff
@@ -18,4 +19,21 @@ import_config:
...
+  runs:
+    - transactions:
+      - output_name: transfer_from_a_to_b
+        script_path: path/to/script1
+        fund_address: address_a
+      - output_name: transfer_from_b_to_c
+        script_path: path/to/script2
+    - transactions:
+      # Note: we've generated transactions for script1, we don't need the name anymore.
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/README.md:32]: in this case, do we just not generate another transaction file even though the transaction still gets submitted through the network?

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/config.rs [hunk 13]
```diff
@@ -101,6 +125,31 @@ pub struct TransactionImporterPerNetworkConfig {
     pub versions_to_import: HashMap<u64, String>,
 }
 
+/// Configuration for generating transactions from scripts.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ScriptTransactionGeneratorConfig {
+    /// List of scripts to run to generate transactions.
+    /// Note, each run will be executed in sequence.
+    pub runs: Vec<ScriptTransactions>,
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/src/config.rs:133]: nit: maybe rename this field to transactions to be more intuitive? wdyt?

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 11 above)
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-transaction-generator/README.md:23]: What is the motivation for having a centralized config for the scripted transactions? Centralized config makes sense for imported transactions, but it's difficult to manage when there's multiple PRs touching this file and results in merge conflicts. 

Another option is to have a central folder where all the scripts live. The transaction generator can pickup code changes in that folder and regen the transactions if needed.

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:23]: done

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 12 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:32]: yes, this is just for dependent transactions. for individual transactions, can be submitted in one run.

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/config.rs (hunk 13 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/src/config.rs:133]: transactions is the subfield of runs; it'll degrade the readabilit.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 11 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:23]: the other option doesn't work since there may be dependent transactions and if we put all dependent transactions under the same folder, we lose the sharable steps.

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:23]: The conflicts are expected for both importing and scripting; this is managable if we append these runs into the config.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 3 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:12]: updated; we have both README and config_template to showcase the usage.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 11 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:23]: One solution can be:
1. move_folder to hold all scripts
2. multiple config files for script transactions. like
```
# fund, transfer and burn.
transactions:
   - name: fund
      path:
   - name: transfer
      path:
   - name: burn
      path: 
```
```
# Create NFT and transfer
transactions:
  - name: create
     path: 
  - name: transfer
     path: 
```

Comment (yuunlimm) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:23]: if we name the move script properly, this seems easier to manage to me.

**Review by yuunlimm — APPROVED** [review]

**Comment by yuunlimm** [comment]
looks like config template got removed

**Comment by larry-aptos** [comment]
> looks like config template got removed

added back with a working command.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md (hunk 11 above)
Comment (rtso) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:25]: qq: the shared move scripts might be using different account addresses, how does it know which address to use in the script? will the addresses be set in config file?

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:25]: no the addresses are hardcoded in the config/move files. 

To allow higher flexibility, we need to add support for addresses to be configured and or flexible script args with which we can pass during runtime. This can be a follow-up task.

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:30]: done.

File: ecosystem/indexer-grpc/indexer-transaction-generator/README.md [hunk 14]
```diff
@@ -4,11 +4,19 @@ This tool is to generate transactions for testing purpose.
...
 ## Usage
 
-`cargo run -- --config example.yaml --output-folder /your_path_to_store_transactions/`
+Under root folder, i.e., `aptos-core`, run
+
+```
+cargo run -p aptos-indexer-transaction-generator -- \
+  --config ecosystem/indexer-grpc/indexer-transaction-generator/template_config.yaml \
+  --output-folder /tmp/ --move-folder-path \
+  ecosystem/indexer-grpc/indexer-transaction-generator/example_scripts
```
Comment (yuunlimm) [inline ecosystem/indexer-grpc/indexer-transaction-generator/README.md:13]: is move-folder-path still required even though we specific the path for the each script in the config?

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:13]: We talked offline; I shared the reason why we determined the current CLI input. 

TL;DR
To get the root of all move folders we can
1. ask users what the folder name is or
2. enforce the name of the root folder or
3. take a guess that something looks like the root of move folders
4. the current way, very explicit on the input. 


I chose the last one since the user needs to specify something anyway or they need to follow some rules.

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-transaction-generator/README.md:13]: Fixed now.

**Review by dermanyang — APPROVED** [review]
approving first; will do post-submission review

File: Cargo.toml [hunk 15]
```diff
@@ -641,7 +641,7 @@ more-asserts = "0.3.0"
 named-lock = "0.2.0"
 native-tls = "0.2.10"
 neptune = { version = "13.0.0", default_features = false }
-ntest = "0.9.3"
+ntest = "0.9.3 "
```
Comment (dermanyang) [inline Cargo.toml:644]: intended?

File: ecosystem/indexer-grpc/indexer-transaction-generator/src/config.rs [hunk 16]
```diff
@@ -25,34 +33,103 @@ pub struct IndexerCliArgs {
...
-pub struct TransactionGeneratorConfig {
-    pub import_config: TransactionImporterConfig, // TODO: Add scripted transaction generation configuration.
-}
+        // Run the script transaction generator.
+        let script_transactions_output_folder =
+            output_folder.join(SCRIPTED_TRANSACTIONS_FOLDER);
+        let move_folder_path = testing_folder.join(MOVE_SCRIPTS_FOLDER);
+        // If the move fixtures folder does not exist, skip the script transaction generator.
+        if !move_folder_path.exists() {
+            return Ok(());
```
Comment (dermanyang) [inline ecosystem/indexer-grpc/indexer-transaction-generator/src/config.rs:68]: do we need a log here?
